core/vote/vote_pool.go

```diff
@@ -0,0 +1,399 @@
+	voteBlockNumber := vote.Data.TargetNumber
+
+	log.Info("The vote info to put is:", "voteBlockNumber", voteBlockNumber, "voteBlockHash", voteBlockHash)
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	if _, ok := m[voteBlockHash]; !ok {
+		// Push into votes priorityQueue if not exist in corresponding votes Map.
+		// To be noted: will not put into priorityQueue if exists in map to avoid duplicate element with the same voteData.
+		heap.Push(votesPq, voteData)
+		voteBox := &VoteBox{
+			blockNumber:  voteBlockNumber,
+			voteMessages: make([]*types.VoteEnvelope, 0, maxCurVoteAmountPerBlock),
+		}
+		m[voteBlockHash] = voteBox
+
+		if isFutureVote {
+			localFutureVotesPqGauge.Inc(1)
+		} else {
+			localCurVotesPqGauge.Inc(1)
+		}
+	}
+
+	// Put into corresponding votes map.
+	m[voteBlockHash].voteMessages = append(m[voteBlockHash].voteMessages, vote)
+	// Add into received vote to avoid future duplicated vote comes.
+	pool.receivedVotes.Add(voteHash)
+	log.Info("VoteHash put into votepool is:", "voteHash", voteHash)
+
+	if isFutureVote {
+		localFutureVotesGauge.Inc(1)
+	} else {
+		localCurVotesGauge.Inc(1)
+	}
+	votesPerBlockHashMetric(voteBlockHash).Inc(1)
+	localReceivedVotesGauge.Inc(1)
+}
+
+func (pool *VotePool) transferVotesFromFutureToCur(latestBlockHeader *types.Header) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	futurePq := pool.futureVotesPq
+	latestBlockNumber := latestBlockHeader.Number.Uint64()
+
+	// For vote before latestBlockHeader-13, transfer to cur if valid.
+	for futurePq.Len() > 0 && futurePq.Peek().TargetNumber+upperLimitOfVoteBlockNumber < latestBlockNumber {
+		blockHash := futurePq.Peek().TargetHash
+		pool.transfer(blockHash)
+	}
+
+	// For vote within latestBlockHeader-13 ~ latestBlockHeader, only transfer the the vote inside the local fork.
+	futurePqBuffer := make([]*types.VoteData, 0)
+	for futurePq.Len() > 0 && futurePq.Peek().TargetNumber <= latestBlockNumber {
+		blockHash := futurePq.Peek().TargetHash
+		header := pool.chain.GetHeaderByHash(blockHash)
+		if header == nil {
+			// Put into pq buffer used for later put again into futurePq
+			futurePqBuffer = append(futurePqBuffer, heap.Pop(futurePq).(*types.VoteData))
+			continue
+		}
+		pool.transfer(blockHash)
+	}
+
+	for _, voteData := range futurePqBuffer {
+		heap.Push(futurePq, voteData)
+	}
+}
+
+func (pool *VotePool) transfer(blockHash common.Hash) {
+	curPq, futurePq := pool.curVotesPq, pool.futureVotesPq
+	curVotes, futureVotes := pool.curVotes, pool.futureVotes
+
+	voteBox := futureVotes[blockHash]
+
+	validVotes := make([]*types.VoteEnvelope, 0, len(voteBox.voteMessages))
+	for _, vote := range voteBox.voteMessages {
+		// Verify if the vote comes from valid validators based on voteAddress (BLSPublicKey).
+		if posa, ok := pool.engine.(consensus.PoSA); ok {
+			if !posa.VerifyVote(pool.chain, vote) {
+				continue
+			}
+		}
+		// Verify the vote from futureVotes.
+		if err := VerifyVoteWithBLS(vote); err == nil {
```

Comment:
This should have done before insert into future.

Reply:
This can not be verified before insert into future, because the header even did not have reached to the range of future votes